### ISSUE TITLE
Transactions failed to be sent to mobile device 

### ISSUE THREAD

**Issue by VexyCats** [WalletConnect/walletconnect-monorepo#374]
Randomly it seems like when a user is prompted to submit a tx - the tx is never sent to their connected device. 

This is hard to replicate as in our dapp it works sometimes, and fails sometimes. Almost like its an internet issue (although its not, as I have a stable fast connection on both devices). 

Replication is hard - but I've noticed: 

If a user directly connects to walletconnect -> accepts in their mobile device -> then uses the dapp to submit a tx it almost never fails. 

But when a user has already connected their wallet, and then they refresh the page or come back to it after a day or two, then once connecting again (not via their device just by clicking on walletconnect in web3modal) there is a good chance the transaction will never appear on their device. 

I've posted in the discord with no real support, but here is the code I'm using for walletconnect and web3modal: 

```
 const providerOptions = {
        walletconnect: {
          package: WalletConnectProvider, // required
          options: {
            infuraId: "a6xxxxxxxxxxxxxxx1", // required
          },
        },
        fortmatic: {
          package: Fortmatic, // required
          options: {
            key: "FORTMATIC_KEY", // required
          },
        },
      };
      const web3Modal = new Web3Modal({
        cacheProvider: false, // optional
        providerOptions, // required
      });
      const provider = await web3Modal.connect();
      store.dispatch({ type: "WEB3_CONNECTED", data: provider });
      this.props.handleCloseWalletModal();
```

The connection seems fine - in terms of when a user refreshes the page and presses walletconnect again - the web3 connection is on the correct network, and is able to create the instance of contract/estimate gas/present the tx data to web3. 

It just fails hits the users device for them to submit, somewhat randomly.

Why would this happen? What aspect of our code/walletconnect would affect this? Is there something we can check/confirm is working that would make the tx fail to be sent to the user after reconnecting from page refresh?

**Comment by miohtama** [comment]
What are dApp, user device and wallet in this case?

**Comment by VexyCats** [comment]
Dapp - Mintable - but testing in localhost (why does this matter?)
User device - android phone 
Wallet: Metamask mobile

**Comment by Bobface** [comment]
I can confirm that I also experience this with my wallet (Argent) everywhere I use WalletConnect (Oasis, Uniswap, Gnosis Safe). It seems that when there is no activity for a longer period the connection breaks. The UIs still show my address and that I am connected, but when I initiate a transaction nothing shows up on my phone while the UI waits for me to confirm. It is then necessary to manually disconnect and connect again using the QR code.

**Comment by VexyCats** [comment]
Hmm... odd, our dapp doesn't have any delays though. As soon as the tx is created its sent to the user via walletconnect.

**Comment by miohtama** [comment]
@VexyCats Some dApps like Gnosis Safe use their own bridge. Also a lot of mobile wallets tend to have issues.

**Comment by miohtama** [comment]
@pedrouid Do you know is there anything special that dApp or wallet needs to do in order to maintain the connected state?

**Comment by VexyCats** [comment]
Yeah I'm unsure what would cause us to lose connection as we use the standard code from the docs and don't take too long - unless its the time from connecting the wallet to submitting the transaction that could be the issue? As we have a user connect their wallet and then it may be 5 minutes before they submit the tx. Although its probably around 30 seconds on average. Shouldn't be that big of an issue right? 


And understood about the dapps, yep, no custom bridge here.

**Comment by sbc64** [comment]
I would like to comment that the current bridge at bridge.walletconnect.org has had a few restart the past couple of days. This is probably the reason as to why a scanning the qr code again is needed. I hope that by today the bridge will be much more stable and that rescanning the qr code won't be as necessary.

Sorry for the reliability issues.

**Comment by miohtama** [comment]
@sebohe and @pedrouid even in the case of a restart, both wallet and dApp should be able find each other again. But maybe I am wrong here. Is there some transient data we lose during the restart?

**Comment by DefiOfThrones** [comment]
Hello guys,

Any update of this problem ?

I'm able to reproduce it any time even on Uniswap.

- Just connect your Metamask wallet on Uniswap with Wallet Connect (On a Android Device) 
- Turn off your internet connection (or your screen during 5 minutes)
- Turn on your internet connection and wait few seconds
- Try to send a transaction with Metamask
- Metamask opens but no transaction confirmation popup appears
- The only way to make it works again is to kill all apps then try to reconnect with Wallet connect several times and it ended up working.

It's a really annoying problem even for the biggest dApps like Uniswap!
This should be a priority in your backlog guys !

**Comment by xwartz** [comment]
> Hello guys,
> 
> Any update of this problem ?
> 
> I'm able to reproduce it any time even on Uniswap.
> 
> * Just connect your Metamask wallet on Uniswap with Wallet Connect (On a Android Device)
> * Turn off your internet connection (or your screen during 5 minutes)
> * Turn on your internet connection and wait few seconds
> * Try to send a transaction with Metamask
> * Metamask opens but no transaction confirmation popup appears
> * The only way to make it works again is to kill all apps then try to reconnect with Wallet connect several times and it ended up working.
> 
> It's a really annoying problem even for the biggest dApps like Uniswap!
> This should be a priority in your backlog guys !

I have the same issue, and my app is React Native.
My guess is that the WS connection was disconnected when the network was disconnected, but the WS was not rebuilt when the network was reconnected.

The code for network monitoring works only in the browser
```js
if (typeof window !== "undefined" && typeof (window as any).addEventListener !== "undefined") {
      window.addEventListener("online", () => this.trigger("online"));
      window.addEventListener("offline", () => this.trigger("offline"));
    }
```
https://github.com/WalletConnect/walletconnect-monorepo/blob/b2cfd26b56/packages/helpers/socket-transport/src/network.ts#L11-L13

**Comment by DefiOfThrones** [comment]
Hey guys, any update on this problem ?

Has anyone found a solution?

**Comment by auuunya** [comment]
+1,
Is there a solution?

**Comment by xwartz** [comment]
I resolved this problem.

1. Add the patch for socket connect error
```
diff --git a/node_modules/@walletconnect/socket-transport/dist/cjs/index.js b/node_modules/@walletconnect/socket-transport/dist/cjs/index.js
index a4fa878..c2e78ff 100644
--- a/node_modules/@walletconnect/socket-transport/dist/cjs/index.js
+++ b/node_modules/@walletconnect/socket-transport/dist/cjs/index.js
@@ -90,6 +90,7 @@ class SocketTransport {
         }
         this._nextSocket.onmessage = (event) => this._socketReceive(event);
         this._nextSocket.onopen = () => this._socketOpen();
+        this._nextSocket.onerror = () => this._socketCreate();
     }
     _socketOpen() {
         this._socketClose();
@@ -147,7 +148,7 @@ class SocketTransport {
             payload: "",
             silent: true,
         }));
-        this._subscriptions = [];
+        // this._subscriptions = [];
     }
     _setToQueue(socketMessage) {
         this._queue.push(socketMessage);

```
2. Reinstantiate the WalletConnect when the network changed from offline to online
```
static handleNetChange(isConnected: boolean) {
      if (isOffline && isConnected) {
        // clear the events
        this.currentInstance.connector._eventManager._eventEmitters = []
        this.currentInstance.connector = null
        this.currentInstance = null
        // reconnecting
        this.restoreSession()
      }
 }
```

**Comment by auuunya** [comment]
> restoreSession

Hey, do you have an example here? May I have a look at it？

**Comment by BenjaminPiette** [comment]
I'm also having this problem with 1.5.4. That version of socket-transport seems to contain a patch that is different than the above patch from @xwartz but that should normally work. I'm going to debug this to understand where the whole mechanism fails to reinit itself.

After turning the screen off/on, we can clearly see that a new socket is created, but it never sends any message. I expect that new socket to send a "sub" request to the relay server, which the code seems to do, but in practice this doesn't happen.

EDIT: Then yes, part of the above patch is missing in current code: this._subscriptions may not be reset to [] after the initial _queueSubscription() otherwise when a new socket connects, there are no subscriptions in the list and no handshake is made.

**Comment by xwartz** [comment]
@BenjaminPiette  yes, the `this._subscriptions` should not be reset to `[]`

**Comment by BenjaminPiette** [comment]
@pedrouid this task can be closed I guess, this is in 1.6.2

**Comment by finessevanes** [comment]
@VexyCats , @DefiOfThrones , @auuunya are you still experiencing the issue?
